File: leetsync/2707-merge-two-2d-arrays-by-summing-values/merge-two-2d-arrays-by-summing-values.go
func mergeArrays(nums1 [][]int, nums2 [][]int) [][]int {
	i, j := 0, 0
	result := [][]int{}
	len1 := len(nums1)
	len2 := len(nums2)
	for i < len1 && j < len2 {
		id1, value1 := nums1[i][0], nums1[i][1]
		id2, value2 := nums2[j][0], nums2[j][1]

		if id1 < id2 {
			result = append(result, []int{id1, value1})
			i++
		} else if id1 > id2 {
			result = append(result, []int{id2, value2})
			j++
		} else {
			result = append(result, []int{id1, value1 + value2})
			i++
			j++
		}
	}
	for i < len1 {
		result = append(result, []int{nums1[i][0], nums1[i][1]})
		i++
	}
	for j < len2 {
		result = append(result, []int{nums2[j][0], nums2[j][1]})
		j++
	}
	return result
}

File: leetsync/2707-merge-two-2d-arrays-by-summing-values/merge-two-2d-arrays-by-summing-values.java

class Solution {
    public int[][] mergeArrays(int[][] nums1, int[][] nums2) {
        List<int[]> resultList = new ArrayList<>();
        int len1 = nums1.length;
        int len2 = nums2.length;
        int i = 0, j = 0;
        while (i < len1 && j < len2) {
            int id1 = nums1[i][0];
            int id2 = nums2[j][0];
            if (id1 == id2) {
                resultList.add(new int[] { id1, nums1[i][1] + nums2[j][1] });
                i++;
                j++;
            } else if (id1 < id2) {
                resultList.add(new int[] { id1, nums1[i][1] });
                i++;
            } else {
                resultList.add(new int[] { id2, nums2[j][1] });
                j++;
            }
        }
        while (i < len1) {
            resultList.add(new int[] { nums1[i][0], nums1[i][1] });
            i++;
        }
        while (j < len2) {
            resultList.add(new int[] { nums2[j][0], nums2[j][1] });
            j++;
        }
        int len = resultList.size();
        int[][] result = new int[len][2];
        for (i = 0; i < len; i++) {
            result[i][0] = resultList.get(i)[0];
            result[i][1] = resultList.get(i)[1];
        }
        return result;
    }
}